### PR TITLE
rust: set sourceProvenance for bootstrap binary

### PR DESCRIPTION
**Pull request by linsui**
## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

@NixOS/rust

### CODE REVIEW

File: pkgs/development/compilers/rust/binary.nix [hunk 1]
```diff
@@ -71,6 +72,7 @@ rec {
 
     meta = with lib; {
       homepage = "http://www.rust-lang.org/";
+      sourceProvenance = with sourceTypes; [ binaryNativeCode ];
       description = "A safe, concurrent, practical language";
```
Comment (linsui) [inline pkgs/development/compilers/rust/binary.nix:76]: This description is also obviously wrong.

Comment (Mic92) [reply pkgs/development/compilers/rust/binary.nix:76]: How so?

Comment (linsui) [reply pkgs/development/compilers/rust/binary.nix:76]: This is cargo, not rust.

Comment (Mic92) [inline pkgs/development/compilers/rust/binary.nix:76]:
```suggestion
      description = "The Rust package manager";
```

File: pkgs/development/compilers/rust/binary.nix [hunk 2]
```diff
@@ -71,6 +72,7 @@ rec {
 
     meta = with lib; {
       homepage = "http://www.rust-lang.org/";
```
Comment (Mic92) [inline pkgs/development/compilers/rust/binary.nix:74]:
```suggestion
      homepage = "https://doc.rust-lang.org/cargo/";
```

File: pkgs/development/compilers/rust/binary.nix (hunk 1 above)
Comment (Mic92) [reply pkgs/development/compilers/rust/binary.nix]: fixed.

File: pkgs/development/compilers/rust/binary.nix [hunk 3]
```diff
@@ -27,6 +27,7 @@ rec {
...
```
Comment (Mic92) [inline pkgs/development/compilers/rust/binary.nix:29]:
```suggestion
      homepage = "https://www.rust-lang.org/";
```

**Review by Mic92 — APPROVED** [review]

File: pkgs/development/compilers/rust/binary.nix (hunk 1 above)
Comment (linsui) [reply pkgs/development/compilers/rust/binary.nix]: Thanks! :)

File: pkgs/development/compilers/rust/binary.nix [hunk 4]
```diff
@@ -26,7 +26,8 @@ rec {
     inherit src;
 
     meta = with lib; {
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/rust/binary.nix:28]: Also why is meta not near the end of the mkDerivation function?

**Comment by SuperSandro2000** [comment]
Also the go boostrapping tarball is also not setting this.

Also I am not a big fan of the because now allowNonSourcePredicate requires a default list of things, otherwise nothing works but I also do not have a good solution to that.

**Comment by linsui** [comment]
> Also the go boostrapping tarball is also not setting this.

For golang, https://github.com/NixOS/nixpkgs/pull/276857. For java, https://github.com/NixOS/nixpkgs/pull/275205.

> Also I am not a big fan of the because now allowNonSourcePredicate requires a default list of things, otherwise nothing works but I also do not have a good solution to that.

This made me confused about why only the bootstrap jdk is there. So l thought it's better to make it clear which packages are not fully bootstrapable yet. I thought go is easy to bootstrap. And hopefully this can encourage people working on this.

File: pkgs/development/compilers/rust/binary.nix (hunk 4 above)
Comment (Mic92) [reply pkgs/development/compilers/rust/binary.nix:28]: no idea.

**Comment by SuperSandro2000** [comment]
> I thought go is easy to bootstrap.

Nope, upstream has no real interest in that either. There standpoint is for a longer while to just download their older artifacts to build the current version. We could go all the way back like rust but that would be very compute intensive.

**Comment by linsui** [comment]
> Nope, upstream has no real interest in that either. There standpoint is for a longer while to just download their older artifacts to build the current version. We could go all the way back like rust but that would be very compute intensive.

I don't understand. Currently only go on musl is built with gccgo. Is there a reason? And it's not as difficult as rust to bootstrap. Go 1.4 is written in C. It can build Go 1.17.3 which can build 1.21.5. The path is short and Go builds very fast.

**Comment by SuperSandro2000** [comment]
> Go 1.4 is written in C. It can build Go 1.17.3 which can build 1.21.5. The path is short and Go builds very fast.

That's interesting. I thought we would need to build every second version. If it it really just those 3 builds then that would probably be viable to bootstrap.

**Comment by linsui** [comment]
I opened an issue for the go bootstrap discussion. https://github.com/NixOS/nixpkgs/issues/277244

**Comment by risicle** [comment]
When we introduced `sourceProvenance` we deliberately didn't mark bootstrap packages as binary because it would cause so much noise like this.

If we're marking bootstrap packages as binary, then technically we should also mark stdenv's bootstrap-tools, and then _everything_ will be disabled for non-allownonsource users.

In the original design I had some ideas around ways of special-casing bootstrap binaries, but I thought people's interest in expanding `sourceProvenance` coverage had reached capacity (See unreviewed PRs like #179845). If people are ready to have that discussion I'd be glad to have it, but in the meantime I think marking packages like this is too disruptive.

**Comment by linsui** [comment]
> If we're marking bootstrap packages as binary, then technically we should also mark stdenv's bootstrap-tools, and then everything will be disabled for non-allownonsource users.

I thought that's good and will be resolved by minimal-bootstrap. Before we get a fully bootstrapable system I thought it's reasonable to tell the user the truth. I noticed this because the jdk used to bootstrap openjdk is also marked.

**Comment by risicle** [comment]
Perhaps. Really I'd like more peoples' thoughts on this.

minimal-bootstrap still won't be zero-binary though will it? And as long as it isn't there will still be an argument that it should be marked as non-source.

My thinking for mechanisms to handle bootstrap packages would be to have a wrapper function of some sort (`forBootstrap` ?) that would override-away the wrapped derivation's `sourceProvenance` and possibly also attempt some assertions that the package isn't being referenced in the output (impossible to make this bulletproof obviously).

**Comment by linsui** [comment]
> minimal-bootstrap still won't be zero-binary though will it? And as long as it isn't there will still be an argument that it should be marked as non-source.

I'm not sure. It's bootstraped from [this file](https://github.com/oriansj/bootstrap-seeds/blob/master/POSIX/x86/hex0_x86.hex0) which is a binary but I thought it's also a source code.

> My thinking for mechanisms to handle bootstrap packages would be to have a wrapper function of some sort (forBootstrap ?) that would override-away the wrapped derivation's sourceProvenance and possibly also attempt some assertions that the package isn't being referenced in the output (impossible to make this bulletproof obviously).

I thought we can add a bootstrapBinary sourceProvenance and convert the sourceProvenance to it in the wrapper.

**Comment by risicle** [comment]
> I thought we can add a bootstrapBinary sourceProvenance and convert the sourceProvenance to it in the wrapper.

Yes that would be neat.

The only issue is that there isn't a _wonderful_ way for users to customize what is considered to be "non-source" at the moment. `allowNonSourcePredicate` is based around making per-_package_ exemptions. It was designed like this to follow the pattern of `allowUnfreePredicate`, which of course has a similar problem.

**Comment by SuperSandro2000** [comment]
> allowNonSourcePredicate is based around making per-package exemptions.

It now basically needs a prefilled list of bootstrap entries if you want to avoid none source packages as much as possible.

**Pull request merged**